src/save_processes.c

```diff
@@ -0,0 +1,540 @@
+#include "global.h"
+#include "ewram.h"
+#include "script.h"
+#include "background.h"
+#include "sound_control.h"
+#include "agb_sram.h"
+
+u32 SaveGameData()
+{
+    gSaveDataBuffer.main.unk17 |= 0x10;
+    DmaCopy16(3, gSaveVersion, gSaveDataBuffer.saveDataVer, 0x30);
+    CalculateSaveChecksum();
+    return WriteSramEx((void*)&gSaveDataBuffer, SRAM_START, sizeof(gSaveDataBuffer));
+}
+
+u32 LoadSaveData()
+{
+    u32 i;
+    char * sramVer;
+    ReadSram(SRAM_START, (void*)&gSaveDataBuffer, 0x29D0);
+    sramVer = gSaveDataBuffer.saveDataVer;
+    for(i = 0; i < 0x30; i++)
+    {
+        if(gSaveVersion[i] != *sramVer)
+        {
+            gMain.unk17 = 0;
+            return 2;
+        }
+        sramVer++;
+    }
+    gMain.unk8E = gSaveDataBuffer.main.unk8E;
+    if(CheckSaveChecksum() == FALSE)
+    {
+        gMain.unk17 = 0;
+        return 1;
+    }
+    gMain.unk17 = 0x10;
+    return 0;
+}
+
+void CalculateSaveChecksum()
+{
+    u8 *saveData;
+    u32 magic;
+    u32 size;
+    gSaveDataBuffer.magic = 0;
+    magic = 0;
+    size = sizeof(gSaveDataBuffer);
+    saveData = (void *) (&gSaveDataBuffer.main);
+    while (saveData < (((u8 *) (&gSaveDataBuffer)) + size))
+    {
+        magic += saveData[0];
+        saveData += 4;
+    }
+    magic += 2343;
+    gSaveDataBuffer.magic = magic;
+}
+
+bool32 CheckSaveChecksum()
+{
+    u32 magic = 0;
+    u8 * saveData = (u8 *)&gSaveDataBuffer.main;
+    while(saveData < (u8 *)&gSaveDataBuffer+sizeof(gSaveDataBuffer))
+    {
+        magic += *saveData;
+        saveData += 4;
+    }
+    magic += 2343;
+    if(magic == gSaveDataBuffer.magic)
+    {
+        return TRUE;
+    }
+    return FALSE;
+}
+
+void ClearSaveProcess(struct Main *main)
+{
+    struct OamAttrs * oam;
+    u32 i;
+    switch (main->process[GAME_SUBPROCESS])
+    {
+    case 0:
+        DmaCopy16(3, gUnknown_08185D20, VRAM + 0x3800, 0x800);
+        DmaCopy16(3, GetBGPalettePtr(0), PLTT, 0x200);
+        DmaCopy16(3, gUnknown_08186540, VRAM, 0x1000);
+        DmaCopy16(3, gUnknown_081964A8, OBJ_VRAM0 + 0x3C00, 0x800);
+        DmaCopy16(3, gUnknown_081FD92C, PLTT + 0x320, 0x40);
+        DmaCopy16(3, gTextPal, PLTT + 0x200, 0x20);
```

Comment:
Could `sizeof` be used for some of these? Or was Capcom weird here

Reply:
i didn't actually have the externs have sizes but i could fix that

Reply:
some stuff is still missing but it's because i can't put concrete sizes on them since capcom preprocessed graphics into a giant binary blob with addresses in defines to access them at an offset